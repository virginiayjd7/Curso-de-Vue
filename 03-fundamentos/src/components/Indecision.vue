<template>
 <img v-if="img" :src="img" alt="bg">
  <div class="bg dark"></div>
<div class="indecision-container">
  <input 
    v-model="question"
    type="text"
    placeholder="Haze una pregunta">
  <p> Recuerda terminar con unsigno de interrogacion(?)</p>
  <div v-if="isValidQuestion">
    <h2>{{question}}</h2>
    <h1>{{answer==='yes'? 'si':'no'}}</h1>
  </div>

</div>
</template>

<script>
export default {
data() {
  return {
    question:null,
    answer:null,
    img:null,
    isValidQuestion:false 
  }
},
methods: {
  async getAnswer(){

    try{
      this.answer='Pensando.....'
      const{answer,image}=await fetch('https://yesno.wtf/api').then(r=>r.json())
      this.answer=answer ==='yes!'? 'Si!':'No!'
      this.img=image

    }catch(error){
      this.answer='No se pudo cargar del API'
      this.img=null

    }
    
  }
},
watch:{
  question(value,oldValue){
    //console.log({value,oldValue})
    this.isValidQuestion=false
    console.log({value})
    //console.log({value})
    if(!value.includes('?'))return  
    //realizar peticion http
    this.isValidQuestion=true
    console.log({value})
    this.getAnswer()
  }
}

}
</script>


<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>
