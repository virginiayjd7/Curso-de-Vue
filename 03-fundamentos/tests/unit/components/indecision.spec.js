import { shallowMount } from '@vue/test-utils'
import Indecision from '@/Components/Indecision.vue'

describe('Indecision Component', () => {
    let wrapper
    let clgSpy

    global.fetch = jest.fn(() => Promise.resolve({
            json: () => Promise.resolve({
                answer: 'yes',
                forced: false,
                image: 'https://yesno.wtf/assets/yes/2.gif'
            })
        }))
        //let getAnswerSpy
    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, 'log')
            //getAnswerSpy = jest.spyOn(console, 'log')
        jest.clearAllMocks()
    })

    test('debe de hacer math con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })
    test('escribir en el input no debe de disparar nada (console.log)', async() => {
        //console.log(wrapper) //es un obejto wrapper
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Hola Mundo')

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()

        //console.log(wrapper.vm)
    })
    test('escribir el simbolo de "?" debe de disparar el getAnswer', async() => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Hola Mundo?')

        expect(clgSpy).toHaveBeenCalledTimes(2)
        expect(getAnswerSpy).toHaveBeenCalled()


    })

    test('pruebas de getAnswer', async() => {
        await wrapper.vm.getAnswer()
            //console.log(wrapper.vm.img)
            //console.log(wrapper.vm.answer)
        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('No!')
    })
    test('pruebas en getAnswer - fallo en el API', async() => {

        //todo fallo en el API
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeFalsy() //nodebe exixtir
        expect(wrapper.vm.answer).toBe('No se pudo cargar del API')
    })

})