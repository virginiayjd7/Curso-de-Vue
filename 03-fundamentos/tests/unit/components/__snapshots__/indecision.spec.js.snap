// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe de hacer math con el snapshot 1`] = `
<!--v-if-->
<div class="bg dark"></div>
<div class="indecision-container"><input type="text" placeholder="Haze una pregunta">
  <p> Recuerda terminar con unsigno de interrogacion(?)</p>
  <!--v-if-->
</div>
`;
